[WIP] CSS: Transitioning from CSS to SCSS.

armaanahluwalia opened the pull request:
This commit begins the transition from using plain CSS to enabling
SCSS compilation using webpack.

* Added the necessary loaders required by webpack to compile SCSS and
import fonts and images.
* Added port forwarding to the Vagrantfile to enable the browser to
access the webpack port from outside the VM.
* Changed the server-side templates to render the portico-styles
bundle instead of directly requiring the portico stylesheet. This
allows webpack to handle stylesheet compilation and minification.
* We use the mini-css-extract-plugin to extract out css from the
includes in webpack and let webpacks production mode handle
minification.
* Moved the hot flag to the webpack-dev-server cli as it handles
adding the plugin and necessary entry points required for hot
code reloading.
* It is now possible to make a change in any of the styles included
by static/styles/scss/portico.scss and see the code reload live
in the browser. The same is true for any of the bundles compiled
by webpack that have the module.accept code in them.

Note: Currently, static/html/5xx.html has a dependency
on the minified portico css code otherwise that entire block can
be removed from the django settings.py.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing Plan:** <!-- How have you tested? -->
Made sure Hot Module Reloading works for both JS and SCSS files including by webpack.
Made sure source maps are correctly displaying for CSS files included by webpack.
Made sure all logged out pages are styled correctly.

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented on Vagrantfile:
```diff
@@ -121,6 +121,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   end
 
   config.vm.network "forwarded_port", guest: 9991, host: host_port, host_ip: host_ip_addr
+  config.vm.network "forwarded_port", guest: 9994, host: 9994, host_ip: host_ip_addr
```
We shouldn't need to do this; if we configure things right, webpack requests should be going to port 9991 and then be being proxied to port 9994 by `tools/run-dev.py`.

armaanahluwalia replied:
As far as I can tell, this is the best possible solution to make HMR work with the Vagrant VM. I tried to forward the calls to sockjs-node using the config file but for some reason my attempts were unsuccessful. Also, take a look at https://bugzilla.mozilla.org/show_bug.cgi?format=default&id=1362443 someone else from mozilla came to the same conclusion i think.

timabbott commented on templates/analytics/stats.html:
```diff
@@ -7,7 +7,7 @@
 
 {% block customhead %}
-{% stylesheet 'portico' %}
+{{ render_bundle('portico-styles') }}
```
It's actually an unrelated bug that we have this stylesheet included directly in so many pages.  I've fixed that by merging 
697fd3c69b1fc8923097f3f152bbd980a44d844d; if you rebase this, you should be able to drop the changes in all but a couple files (since those no longer directly include the portico stylesheet).

armaanahluwalia replied:
Rebased

timabbott commented on tools/webpack:
```diff
@@ -29,7 +29,18 @@ def run_watch(host, port, minify):
+        '--host',
+        host,
+        '--port',
+        port,
```
For readability, I would probably use one line each for pairs of options with a key and value like `host` and `port`.

timabbott commented on tools/webpack.assets.json:
```diff
@@ -31,5 +31,6 @@
     "translations": "./static/js/translations.js",
-    "zxcvbn": "./node_modules/zxcvbn/dist/zxcvbn.js"
+    "zxcvbn": "./node_modules/zxcvbn/dist/zxcvbn.js",
+    "portico-styles": "./static/styles/scss/portico.scss"
```
Is there a useful way to put the CSS part of this into its own file with a separate namespace?  Or do we need to carefully name all the CSS files with `-styles` is in the name to avoid having the JS and CSS bundles with a given name have the same name?  

(Or is the normal thing to do to make the `portico` bundle have both JS and CSS in it?)

armaanahluwalia replied:
The ideal scenario is that we import scss files from the associated JS files for the page ( which ideally there would be one per page ). If there are multiple JS files being included in the page we can create a container file that imports those other files and the CSS files. For this PR I opted to add a portico-styles bundle separately because I didn't want to touch JS files associated with all the various places the portico stylesheet was being imported. I think when we decide to go forward we should take the import CSS from JS approach and move this portico file to an associated JS file as well.

timabbott commented on tools/webpack.config.ts:
```diff
@@ -1,9 +1,27 @@
 
+// Currently we're using style-loader for local dev
+// because MiniCssExtractPlugin doesn't support
+// HMR as yet. When this changes we can switch
```
Is there an issue in MiniCssExtractPlugin for this feature that we can link to?

armaanahluwalia replied:
Added link to issue.

timabbott commented on tools/webpack.config.ts:
```diff
@@ -62,14 +80,59 @@ export default (env?: string) : webpack.Configuration => {
+                            name: '[name].[ext]',
+                            outputPath: 'fonts/'
+                        }
+                    }]
```
Just to make sure I'm reading this correctly, block code is unused for now, right?

armaanahluwalia replied:
This is being used by portico.scss which imports fonts via fonts.css

timabbott commented on tools/webpack.config.ts:
```diff
@@ -81,6 +144,11 @@ export default (env?: string) : webpack.Configuration => {
             new BundleTracker({filename: 'webpack-stats-production.json'}),
+            // Extract CSS from files
+            new MiniCssExtractPlugin({
+                filename: "[name][hash].css",
```
We should use a `[name].[hash].css` naming scheme with a clear delimiter between the filename and the hash, right?

priyankp10 replied:
We should start using  `[contenthash]`. I think I works better.

timabbott replied:
@priyankp10 can you explain your reasoning?

priyankp10 replied:
The content hash is the hash generated from the text content it consumed, so it only invalidates the hash if the file is changed, rather than the whole thing.

I am try to explain it the best of it as i can, it hash to do with how webpack's entry point or our dependencies (in this case using `@import`) and hash are generated  -- contenthash does the better job for it.

To summarize, just provides better hashing.

(I look online for something that can explain this, [this article](https://survivejs.com/webpack/optimizing/adding-hashes-to-filenames/) does that pretty well:  )

timabbott commented on tools/webpack.config.ts:
```diff
@@ -81,6 +144,11 @@ export default (env?: string) : webpack.Configuration => {
+            new MiniCssExtractPlugin({
+                filename: "[name][hash].css",
+                chunkFilename: "[id].css"
+            })
```
Have you tested this production configuration on a production machine?

timabbott commented on tools/webpack.config.ts:
```diff
@@ -91,21 +159,21 @@ export default (env?: string) : webpack.Configuration => {
-                poll: 100,
-            },
+                poll: 100
+            }
```
What's the story with the changes in options for the dev server?  This seems unrelated and I'm surprised you didn't do a separate commit for them...

armaanahluwalia replied:
Moved to separate commit.

timabbott commented on package.json:
```diff
@@ -25,15 +27,19 @@
     "string.prototype.codepointat": "0.2.0",
     "string.prototype.endswith": "0.2.0",
     "string.prototype.startswith": "0.2.0",
+    "style-loader": "0.21.0",
```
I think you could do the `webpack` and `yarn.lock` changes in a separate first commit, so that we can merge those (since they shouldn't break anything) while we iterate on the rest of the configuration.

armaanahluwalia replied:
Moved to separate commit.

priyankp10 commented on tools/webpack.config.ts:
```diff
@@ -1,9 +1,27 @@
+// over to MiniCssExtractPlugin which will Also
+// solve the flash of unstsyled elements on page
+// load
+const getCSSLoader = (isProd) => {
```
This should be a function and should be type annotated.
```typescript
function getCSSLoader(isProd: boolean) {
```

armaanahluwalia commented:
@priyankp10 Thanks priyank for your comments on this PR. I fixed the typescript function. Regarding contentHash you're referring to an option for a different plugin ( extract-text-webpack-plugin ) which we're not using here.

@timabbott I've fixed all issues except the Vagrantfile as discussed with you. For some reason the build is failing I'm trying to understand why. Any insight there would be really helpful.

priyankp10 commented:
@armaanahluwalia `[contenthash]` should work for `mini-css-extract-plugin`, i tried it out locally  to confirm and the readme actually suggest using `contenthash`: https://github.com/webpack-contrib/mini-css-extract-plugin#long-term-caching

armaanahluwalia commented:
@priyankp10 Ahh interesting totally missed that. Thanks!

timabbott commented:
I fixed the test failure, cleaned up a few small things (most notably that `portico-help.html` hadn't been updated, but also making the new hot number `host_port + 3`, not the hardcoded 9994), tested it in production on chat.zulip.org and merged this, thanks @armaanahluwalia!
